File: frontend/src/api/musicEngine.ts
import apiClient from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface MusicJob {
  id: string
  project_id: string
  scene_id: string | null
  episode_id: string | null
  job_type: string
  status: string
  mood: string
  triggered_by: string
  params: Record<string, unknown>
  result: Record<string, unknown> | null
  error_message: string | null
  started_at: string | null
  completed_at: string | null
  created_at: string
  updated_at: string
}

export interface MusicOutput {
  id: string
  job_id: string
  project_id: string
  scene_id: string | null
  episode_id: string | null
  output_type: string
  mood: string
  loop_type: string
  storage_key: string
  duration_seconds: number
  sample_rate: number
  format: string
  file_size_bytes: number
  provider: string
  copyright_safe: boolean
  status: string
  created_at: string
}

export interface SFXAsset {
  id: string
  sfx_key: string
  name: string
  description: string
  category: string
  tags: string[]
  storage_key: string
  duration_seconds: number
  format: string
  sample_rate: number
  is_builtin: boolean
  is_active: boolean
  created_at: string
}

export interface MusicRetryEntry {
  id: string
  project_id: string
  scene_id: string | null
  episode_id: string | null
  original_job_id: string | null
  status: string
  retry_count: number
  max_retries: number
  reason: string
  params: Record<string, unknown>
  next_retry_at: string | null
  created_at: string
  updated_at: string
}

export interface PaginationMeta {
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface MusicJobListResponse {
  items: MusicJob[]
  meta: PaginationMeta
}

export interface MusicOutputListResponse {
  items: MusicOutput[]
  meta: PaginationMeta
}

export interface SFXAssetListResponse {
  items: SFXAsset[]
  meta: PaginationMeta
}

export interface MusicRetryQueueListResponse {
  items: MusicRetryEntry[]
  meta: PaginationMeta
}

export interface MusicDashboardStats {
  total_jobs: number
  jobs_completed: number
  jobs_pending: number
  jobs_failed: number
  jobs_running: number
  total_music_outputs: number
  total_sfx_assets: number
  total_retry_queue: number
  recent_jobs: MusicJob[]
}

export interface DispatchResponse {
  job_id: string
  status: string
  message: string
  dispatch_mode: string
}

export interface TriggerMusicTrackRequest {
  project_id: string
  scene_id?: string
  episode_id?: string
  mood?: string
  duration_seconds?: number
  loop_type?: string
  prompt?: string
  bpm?: number
  instruments?: string[]
  output_format?: string
  extra_params?: Record<string, unknown>
}

export interface TriggerSceneAudioRequest {
  project_id: string
  scene_id: string
  episode_id?: string
  mood?: string
  duration_seconds?: number
  output_format?: string
  include_sfx?: boolean
  sfx_keys?: string[]
  extra_params?: Record<string, unknown>
}

// ---------------------------------------------------------------------------
// API client
// ---------------------------------------------------------------------------

export const musicEngineApi = {
  getDashboard: (projectId: string): Promise<MusicDashboardStats> =>
    apiClient.get(`/mu/dashboard/${projectId}`).then((r) => r.data),

  // Jobs
  listJobs: (
    projectId: string,
    params?: { page?: number; status?: string; job_type?: string }
  ): Promise<MusicJobListResponse> =>
    apiClient
      .get('/mu/jobs', { params: { project_id: projectId, ...params } })
      .then((r) => r.data),

  getJob: (jobId: string): Promise<MusicJob> =>
    apiClient.get(`/mu/jobs/${jobId}`).then((r) => r.data),

  // Outputs
  listOutputs: (
    projectId: string,
    params?: { page?: number; mood?: string; output_type?: string; status?: string }
  ): Promise<MusicOutputListResponse> =>
    apiClient
      .get('/mu/outputs', { params: { project_id: projectId, ...params } })
      .then((r) => r.data),

  getOutput: (outputId: string): Promise<MusicOutput> =>
    apiClient.get(`/mu/outputs/${outputId}`).then((r) => r.data),

  // SFX Library
  listSFX: (params?: {
    page?: number
    category?: string
    search?: string
  }): Promise<SFXAssetListResponse> =>
    apiClient.get('/mu/sfx', { params }).then((r) => r.data),

  getSFX: (sfxKey: string): Promise<SFXAsset> =>
    apiClient.get(`/mu/sfx/${sfxKey}`).then((r) => r.data),

  // Retry queue
  listRetryQueue: (
    projectId: string,
    params?: { page?: number; status?: string }
  ): Promise<MusicRetryQueueListResponse> =>
    apiClient
      .get('/mu/retry-queue', { params: { project_id: projectId, ...params } })
      .then((r) => r.data),

  processRetryQueue: (projectId: string): Promise<DispatchResponse> =>
    apiClient
      .post('/mu/retry-queue/process', null, { params: { project_id: projectId } })
      .then((r) => r.data),

  // Triggers
  generateTrack: (body: TriggerMusicTrackRequest): Promise<DispatchResponse> =>
    apiClient.post('/mu/generate/track', body).then((r) => r.data),

  generateSceneAudio: (body: TriggerSceneAudioRequest): Promise<DispatchResponse> =>
    apiClient.post('/mu/generate/scene-audio', body).then((r) => r.data),
}

File: backend/apps/api/schemas/music_engine.py
"""
Phase 9 — Music & Sound Engine Pydantic v2 schemas.
Mirrors animation_engine.py / voice_engine.py schemas shape exactly.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Shared
# ---------------------------------------------------------------------------

class PaginationMeta(BaseModel):
    total: int
    page: int
    page_size: int
    total_pages: int


# ---------------------------------------------------------------------------
# MusicGenerationJob
# ---------------------------------------------------------------------------

class MusicJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    scene_id: UUID | None = None
    episode_id: UUID | None = None
    job_type: str
    status: str
    mood: str
    triggered_by: str
    params: dict[str, Any]
    result: dict[str, Any] | None = None
    error_message: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class MusicJobListResponse(BaseModel):
    items: list[MusicJobResponse]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# MusicOutput
# ---------------------------------------------------------------------------

class MusicOutputResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    job_id: UUID
    project_id: UUID
    scene_id: UUID | None = None
    episode_id: UUID | None = None
    output_type: str
    mood: str
    loop_type: str
    storage_key: str
    duration_seconds: float
    sample_rate: int
    format: str
    file_size_bytes: int
    provider: str
    copyright_safe: bool
    status: str
    created_at: datetime


class MusicOutputListResponse(BaseModel):
    items: list[MusicOutputResponse]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# SoundEffectAsset
# ---------------------------------------------------------------------------

class SFXAssetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    sfx_key: str
    name: str
    description: str
    category: str
    tags: list[str]
    storage_key: str
    duration_seconds: float
    format: str
    sample_rate: int
    is_builtin: bool
    is_active: bool
    created_at: datetime


class SFXAssetListResponse(BaseModel):
    items: list[SFXAssetResponse]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# MusicRetryQueue
# ---------------------------------------------------------------------------

class MusicRetryQueueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    scene_id: UUID | None = None
    episode_id: UUID | None = None
    original_job_id: UUID | None = None
    status: str
    retry_count: int
    max_retries: int
    reason: str
    params: dict[str, Any]
    next_retry_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class MusicRetryQueueListResponse(BaseModel):
    items: list[MusicRetryQueueResponse]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Dashboard
# ---------------------------------------------------------------------------

class MusicDashboardStats(BaseModel):
    total_jobs: int
    jobs_completed: int
    jobs_pending: int
    jobs_failed: int
    jobs_running: int
    total_music_outputs: int
    total_sfx_assets: int
    total_retry_queue: int
    recent_jobs: list[MusicJobResponse]


# ---------------------------------------------------------------------------
# Dispatch / trigger
# ---------------------------------------------------------------------------

class DispatchResponse(BaseModel):
    job_id: UUID
    status: str
    message: str
    dispatch_mode: str


class TriggerMusicTrackRequest(BaseModel):
    """Generate a background music track for a scene or episode."""
    project_id: UUID
    scene_id: UUID | None = None
    episode_id: UUID | None = None
    mood: str = "neutral"       # comedy | adventure | sad | happy | tension | victory | neutral
    duration_seconds: float = 30.0
    loop_type: str = "looping"  # looping | one_shot
    prompt: str = ""
    bpm: int = 0
    instruments: list[str] = []
    output_format: str = "wav"
    extra_params: dict[str, Any] = {}


class TriggerSceneAudioRequest(BaseModel):
    """Generate complete audio (music + SFX) for a scene."""
    project_id: UUID
    scene_id: UUID
    episode_id: UUID | None = None
    mood: str = "neutral"
    duration_seconds: float = 30.0
    output_format: str = "wav"
    include_sfx: bool = True
    sfx_keys: list[str] = []     # specific SFX to mix in
    extra_params: dict[str, Any] = {}
